lib/beaker/puppet_install_helper.rb

```diff
@@ -83,9 +83,13 @@ def run_puppet_install_helper_on(hosts, type_arg = find_install_type, version =
         end
       end
     when 'agent'
-      if find_agent_sha.nil?
+      if ENV['BEAKER_PUPPET_COLLECTION'] =~ /-nightly$/
         # Workaround for RE-10734
-        options[:release_apt_repo_url] = "http://apt.puppetlabs.com/#{ENV['BEAKER_PUPPET_COLLECTION']}" if ENV['BEAKER_PUPPET_COLLECTION'] =~ /-nightly$/
+        options[:release_apt_repo_url] = "http://apt.puppetlabs.com/#{ENV['BEAKER_PUPPET_COLLECTION']}"
+        options[:win_download_url] = 'http://nightlies.puppet.com/downloads/windows'
+        options[:mac_download_url] = 'http://nightlies.puppet.com/downloads/mac'
+      end
+      if find_agent_sha.nil?
         install_puppet_agent_on(hosts, options.merge(version: version))
```

Comment:
I still haven't quite figured out which beaker repo this method is defined in. Same goes for `install_puppet_agent_dev_repo_on`. I think beaker-puppet might have one of these? Any reason not to hit the new puppet5 install methods for development packages from builds.delivery.puppetlabs.net? Are you trying not to hit internal only servers for these jobs for any reason?

Reply:
https://github.com/puppetlabs/beaker-puppet/blob/master/lib/beaker-puppet/install_utils/puppet5.rb#L160 works for anything on builds.delivery.puppetlabs.net (even older versions before puppet5, as long as they've been shipped with automation we added about a year ago).

https://github.com/puppetlabs/beaker-puppet/blob/master/lib/beaker-puppet/install_utils/foss_utils.rb#L927 will install release repos (so deb/rpm based only), but works for either nightlies or final releases. I'm not sure if that method works for anything internal. You have to have the `host.install (package)` call after you enable the repos though.

I don't think we ever got around to adding nightlies for windows/mac, but I bet RE would help with that

Reply:
Yeah, beaker-puppet for all the puppet-y stuff, minus the defaults that I linked above :P. The puppet 5 install methods are not generic and I'm trying to use the most generic install methods possible. builds.delivery.puppetlabs.net is not publicly accessible so community users wouldn't be able to use it. Also to install from that we need to pass the exact version or SHA rather than just grab the latest version from a given collection.

install_puppet_agent_on: https://github.com/puppetlabs/beaker-puppet/blob/63ea32a0d7caa8f261c533b020625de19569f971/lib/beaker-puppet/install_utils/foss_utils.rb#L340

install_puppet_agent_dev_repo_on: https://github.com/puppetlabs/beaker-puppet/blob/63ea32a0d7caa8f261c533b020625de19569f971/lib/beaker-puppet/install_utils/foss_utils.rb#L1140

Reply:
@melissa 
> The puppet5 install method I pointed you to is totally generic!!

ah sorry, I was responding to the first message before the 2nd had come in about the puppet 5 method and so didn't see it.

> You have to have the `host.install(package)` call after you enable the repos though.

Looking at the BPIH code that calls `install_puppet_agent_dev_repo_on()` it is only triggered when `BEAKER_PUPPET_AGENT_SHA` is set, and afaik we never actually use that... and it wouldn't work anyway since as you mention `host.install(package)` [isn't even called](https://github.com/hunner/beaker-puppet_install_helper/blob/140d989341907c7ce363e0f71bbd4bd28b9f5b62/lib/beaker/puppet_install_helper.rb#L97). Maybe I should just delete that code since we don't use it and it's never worked.

> https://github.com/puppetlabs/beaker-puppet/blob/master/lib/beaker-puppet/install_utils/foss_utils.rb#L927 will install release repos (so deb/rpm based only), but works for either nightlies or final releases. I'm not sure if that method works for anything internal. You have to have the host.install (package) call after you enable the repos though.

This is essentially what `install_puppet_agent_on()` does: https://github.com/puppetlabs/beaker-puppet/blob/63ea32a0d7caa8f261c533b020625de19569f971/lib/beaker-puppet/install_utils/foss_utils.rb#L372-L373

> I don't think we ever got around to adding nightlies for windows/mac, but I bet RE would help with that

This PR makes that work :).

> What kind of generic install methods are you looking to have?

My overarching goal for BPIH is that anyone doing acceptance testing should be able to call `run_puppet_install_helper()` in their `spec/spec_helper_acceptance.rb` and have masters/agents set up based on their nodeset type of `pe` or `agent`, nodeset roles of `agent` getting puppet-agent and `master` getting either PE for pe nodesets or puppetserver for non-pe nodesets, plus the handful of related environment variables used by beaker such as `BEAKER_IS_PE`, `BEAKER_PUPPET_COLLECTION`, `BEAKER_PUPPET_AGENT_VERSION`, and `BEAKER_PE_VER`. Much of the code in BPIH was from the old days when setting up puppet 3.x, 4.x, or PE 3.x or 2015.x was complicated. Nowadays `install_pe_on()` and `install_puppet_agent_on()` provide pretty much all of the generic abstraction I need.